cli/internal/oauth/manager/manager.go

```diff
@@ -0,0 +1,234 @@
+	var tokenRes api.TokenResponse
+	select {
+	case <-ctx.Done():
+		return res, errors.New("login canceled")
+	case err := <-waitForTokenErrChan:
+		return res, fmt.Errorf("login failed: %w", err)
+	case tokenRes = <-tokenResChan:
+	}
+
+	claims, err := util.GetClaims(tokenRes.AccessToken)
+	if err != nil {
+		return res, fmt.Errorf("login failed: %w", err)
+	}
+
+	res.Tenant = m.tenant
+	res.AccessToken = tokenRes.AccessToken
+	res.RefreshToken = tokenRes.RefreshToken
+	res.Claims = claims
+
+	if err = m.storeTokensInStore(tokenRes.AccessToken, tokenRes.RefreshToken); err != nil {
+		return res, fmt.Errorf("login failed: %w", err)
+	}
+	return res, nil
+}
+
+// Logout logs out of the session for the client and removes tokens from the storage provider.
+func (m *OAuthManager) Logout() error {
+	return errors.Join(
+		m.eraseTokensFromStore(),
+		m.openBrowser(m.api.LogoutURL()),
+	)
+}
+
+var (
+	// ErrNoCreds is returned by RefreshToken when the store does not contain credentials
+	// for the official registry.
+	ErrNoCreds = errors.New("no credentials found")
+
+	// ErrUnexpiredToken is returned by RefreshToken when the token is not expired.
+	ErrUnexpiredToken = errors.New("token is not expired")
+)
+
+const minimumTokenLifetime = 20 * time.Minute
+
+// RefreshToken fetches credentials from the store, refreshes them, stores the new tokens
+// and returns them.
+// If there are no credentials in the store, ErrNoCreds is returned.
+func (m OAuthManager) RefreshToken() (res oauth.TokenResult, err error) {
+	access, refresh, err := m.fetchTokensFromStore()
+	if err != nil {
+		return res, err
+	}
+	if access == "" {
+		return res, ErrNoCreds
+	}
+
+	claims, err := util.GetClaims(access)
+	if err != nil {
+		return res, err
+	}
+	if claims.Expiry.Time().After(time.Now().Add(minimumTokenLifetime)) {
+		return res, ErrUnexpiredToken
+	}
+
+	refreshRes, err := m.api.Refresh(refresh)
+	if err != nil {
+		return res, err
+	}
+
+	err = m.storeTokensInStore(refreshRes.AccessToken, refreshRes.RefreshToken)
+	if err != nil {
+		return res, err
+	}
+
+	claims, err = util.GetClaims(refreshRes.AccessToken)
+	if err != nil {
+		return res, err
+	}
+
+	res.Tenant = m.tenant
+	res.AccessToken = refreshRes.AccessToken
+	res.RefreshToken = refreshRes.RefreshToken
+	res.Claims = claims
+	return res, nil
+}
+
+const (
+	registryAuthKey = "https://index.docker.io/v1/"
+	accessTokenKey  = "access-token"
+	refreshTokenKey = "refresh-token"
+)
+
+func (m *OAuthManager) fetchTokensFromStore() (access, refresh string, err error) {
+	accessAuth, err := m.credStore.Get(registryAuthKey + accessTokenKey)
+	if err != nil {
+		return access, refresh, fmt.Errorf("failed to fetch access token: %w", err)
+	}
+	access = accessAuth.Password
+
+	refreshAuth, err := m.credStore.Get(registryAuthKey + refreshTokenKey)
+	if err != nil {
+		return access, refresh, fmt.Errorf("failed to fetch refresh token: %w", err)
+	}
+	refresh = refreshAuth.Password
+
+	return
+}
```

Comment:
We _may_ need some other way to construct the access and refresh token keys, or make sure that there's no code that takes the registry hostname into account - like https://github.com/docker/cli/pull/3599 + https://github.com/docker/cli/pull/5195.
Otherwise, the access/refresh token _could_ end up being used as an actual auth token.

I haven't investigated it in detail, so it might not be an actual issue in this case, but thought it's worth bringing up 😅

Reply:
Hmm, that code only does that for non-official registries, so it should be fine (and indeed, that works). We can make sure it continues to work by adding some e2e tests, but maybe we can think of some other way to store these.

I still also want to try to implement some of these at the credstore level, so I'll see how that looks too/if this still matters.

Reply:
Yeah while it might not be a problem for the CLI code itself, I'm curious if other credential helpers might also do something like this.

Implementing these extra tokens on the credential store level would be nice indeed.

Reply:
yeah, I have a working branch now! I think it looks nicer, I'll ask y'all to take a look soon.

Reply:
We can check! What I was going to do was still at the CLI level, but as a wrapper to the CLI "credential store" implementations.